binary/src/Pos/Binary/Class/Core.hs

```diff
@@ -659,3 +760,252 @@ instance (i ~ G.C, GSerialiseProd f) => GSerialiseSum (G.M1 i c f) where
     fieldsForCon _ _ = cborError "Bad constructor number"
     decodeSum      0 = G.M1 <$> gdecodeSeq
     decodeSum      _ = cborError "bad constructor number"
+
+
+----------------------------------------------------------------------------
+-- Size expressions
+----------------------------------------------------------------------------
+
+(.:) :: (c -> d) -> (a -> b -> c) -> (a -> b -> d)
+f .: g = \x y -> f (g x y)
+
+-- | Expressions describing the statically-computed size bounds on
+--   a type's possible values.
+type Size = Fix SizeF
+
+-- | The base functor for @Size@ expressions.
+data SizeF t
+    = AddF t t -- ^ Sum of two sizes.
+    | MulF t t -- ^ Product of two sizes.
+    | SubF t t -- ^ Difference of two sizes.
+    | AbsF t   -- ^ Absolute value of a size.
+    | NegF t   -- ^ Negation of a size.
+    | SgnF t   -- ^ Signum of a size.
+    | CasesF [Case t] -- ^ Case-selection for sizes. Used for sum types.
+    | ValueF Byte     -- ^ A constant value.
+    | ApF String (Byte -> Byte) t
+      -- ^ Application of a monotonic function to a size.
+    | forall a. Bi a => TodoF (forall x. Bi x => Proxy x -> Size) (Proxy a)
```

Comment:
I'm not sure I understand this `TodoF` constructor --- that is, I don't quite understand why it's needed and how it's used. I think this needs a bit more of an explanation.